rclcpp/src/rclcpp/parameter_events_subscriber.cpp

```diff
@@ -0,0 +1,157 @@
+  qos_(qos)
+{}
+
+void ParameterEventsSubscriber::add_namespace_event_subscriber(const std::string & node_namespace)
+{
+  if (std::find(node_namespaces_.begin(), node_namespaces_.end(),
+    node_namespace) == node_namespaces_.end())
+  {
+    node_namespaces_.push_back(node_namespace);
+    auto topic = join_path(node_namespace, "parameter_events");
+    RCLCPP_DEBUG(node_logging_->get_logger(), "Subscribing to topic: %s", topic.c_str());
+
+    auto event_sub = rclcpp::create_subscription<rcl_interfaces::msg::ParameterEvent>(
+      node_topics_, topic, qos_,
+      std::bind(&ParameterEventsSubscriber::event_callback, this, std::placeholders::_1));
+
+    event_subscriptions_.push_back(event_sub);
+  }
+}
+
+void ParameterEventsSubscriber::set_event_callback(
+  std::function<void(const rcl_interfaces::msg::ParameterEvent::SharedPtr &)> callback,
+  const std::string & node_namespace)
+{
+  std::string full_namespace;
+  if (node_namespace == "") {
+    full_namespace = node_base_->get_namespace();
+  }
+
+  full_namespace = resolve_path(full_namespace);
+  add_namespace_event_subscriber(full_namespace);
+  user_callback_ = callback;
+}
+
+void ParameterEventsSubscriber::register_parameter_callback(
+  const std::string & parameter_name,
+  std::function<void(const rclcpp::Parameter &)> callback,
+  const std::string & node_name)
+{
+  auto full_node_name = resolve_path(node_name);
+  add_namespace_event_subscriber(split_path(full_node_name).first);
+  parameter_node_map_[parameter_name] = full_node_name;
+  parameter_callbacks_[parameter_name] = callback;
+}
+
+bool ParameterEventsSubscriber::get_parameter_from_event(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event,
+  rclcpp::Parameter & parameter,
+  const std::string parameter_name,
+  const std::string node_name)
+{
+  if (event->node == resolve_path(node_name)) {
+    rclcpp::ParameterEventsFilter filter(event, {parameter_name},
+      {rclcpp::ParameterEventsFilter::EventType::NEW,
+        rclcpp::ParameterEventsFilter::EventType::CHANGED});
+    if (!filter.get_events().empty()) {
+      auto param_msg = filter.get_events()[0].second;
+      parameter = rclcpp::Parameter::from_parameter_msg(*param_msg);
+      return true;
+    }
+  }
+  return false;
+}
+
+void ParameterEventsSubscriber::event_callback(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
+{
+  std::string node_name = event->node;
+  RCLCPP_DEBUG(node_logging_->get_logger(), "Parameter event received for node: %s",
+    node_name.c_str());
+
+  for (std::map<std::string, std::string>::iterator it = parameter_node_map_.begin();
+    it != parameter_node_map_.end(); ++it)
+  {
+    rclcpp::Parameter p;
+    if (get_parameter_from_event(event, p, it->first, it->second)) {
+      parameter_callbacks_[it->first](p);
+    }
+  }
+
+  if (user_callback_) {
+    user_callback_(event);
+  }
+}
+
+std::string ParameterEventsSubscriber::resolve_path(const std::string & path)
```

Comment:
I feel like some of the functionality in `resolve_path`, 'split_path`, `join_path` should be already provided in `rcl`.
We will probably need to do something similar in other places (if we're not already doing it).

Reply:
I see there is a split string functionality in rcutils: https://github.com/ros2/rcutils/blob/master/include/rcutils/split.h, and I also see there is `rclcpp::expand_topic_or_service_name`. The closest to `resolve_path` I see is `rclcpp::Node::create_effective_namespace` but that is not a public function.

Reply:
And we also have https://github.com/ros2/rcpputils/blob/a5a973c2f3cf403af55e312d0054dc889995ba10/include/rcpputils/join.hpp.

If you can centralize some of this functions somewhere it would be good (some common string functionality in `rcutils` or `rcpputils`, and some specific functions for node names, etc in `rcl`).

But I think that's outside the scope of this PR, so you can left the methods here if you want and I will open an issue for tracking the enhancement.

Reply:
Neat, didn't know about the `join`. I could get rid of the `split_path` and `join_path` using those utils.